### PR TITLE
Add support for FIDO U2F

### LINKED ISSUE #1024: Support FIDO U2F over USB and NFC.
## Why you want U2F

* It authenticates a user via public key authentication between server and client. Stronger encryption than OTP.
* Public key authentication prevents phishing and man-in-the-middle attacks.
* There is no keyboard input, so it is not susceptible to keyboard loggers.
* A U2F device **_can_** generate a new key pair for each (web) application. Automatic generation of application-specific keys prevents websites from tracking you by the same public key.

## Desktop (U2F over USB)

* Already implemented on chrome and opera.
* Available as an addon in firefox. U2F will probably be natively supported in firefox soon.
* According to https://developer.microsoft.com/en-us/microsoft-edge/platform/status/fido20webapis/, Microsoft hasn't started working on U2F, yet.

## Mobile

U2F over NFC is currently supported via google authenticator on android.
However, U2F over USB is still possible if you connect a U2F key device to a smartphone via a USB cable.

## Supported Web Services

* Google
* Facebook
* Github
* GitLab
* Etc
  * http://www.dongleauth.info/
  * https://twofactorauth.org/

### PR DESCRIPTION
**Pull request by jonasfranz**
Fixes #1024 

Adds support for FIDO U2F as an addition to Two-Factor Authentication by Phone. ~~Currently it is only works with Chrome but I am trying to support Firefox and Android too.~~

## Requirements 
* FIDO U2F certified key
* Access via https

## Video example
[gitea fido u2f demo](http://www.youtube.com/watch?v=sA4Id3Sc72M)

## TODO
- [x] Improve error handling
- [x] Android (Chrome) support
- [x] ~~iOS support (Hardware required)~~ *(no software support by iOS currently)*
- [x] Firefox support auth
- [x] Firefox support register
- [x] Expiration timer
- [x] Redirect / Check for https
- [x] Add tests (lots of tests)

## iOS
~~I cannot test iOS at the moment because my security key does not support Bluetooth LE. If you want to provide me a BLE key, please contact me via Discord.~~ Thanks to @techknowlogick for sponsoring a Bluetooth LE key.

### CODE REVIEW

**Comment by jonasfranz** [comment]
Test instance: https://try.h.jonasfranz.software/

**Comment by techknowlogick** [comment]
Firefox support above is checked off however I'm using FF 60.0 (64-bit), on MacOS 10.13.4, and I receive the following message:
screen shot 2018-05-17 at 9 21 06 pm

The key I'm using is: https://www.yubico.com/product/yubikey-neo/ 

Is FF support just FF Mobile?

(I see a similar message when trying to add this key to GitHub, so it is likely my browser just has issues with U2F)

File: custom/conf/app.ini.sample [hunk 1]
```diff
@@ -570,6 +570,14 @@ MAX_RESPONSE_ITEMS = 50
 LANGS = en-US,zh-CN,zh-HK,zh-TW,de-DE,fr-FR,nl-NL,lv-LV,ru-RU,ja-JP,es-ES,pt-BR,pl-PL,bg-BG,it-IT,fi-FI,tr-TR,cs-CZ,sr-SP,sv-SE,ko-KR
 NAMES = English,简体中文,繁體中文（香港）,繁體中文（台灣）,Deutsch,français,Nederlands,latviešu,русский,日本語,español,português do Brasil,polski,български,italiano,suomi,Türkçe,čeština,српски,svenska,한국어
 
+[U2F]
```
Comment (techknowlogick) [inline custom/conf/app.ini.sample:573]: Could you add a U2F section to the "Config Cheatsheet" page in the docs?

**Comment by jonasfranz** [comment]
@techknowlogick Do you enabled u2f via about:config ?
Firefox Mobile for Android does not work at the moment because Google Authenticator only supports Chrome Mobile and not Firefox. U2F is handled by Google Authenticator at the moment. If you enable U2F via about:config on Firefox Mobile it does not work too.

File: custom/conf/app.ini.sample (hunk 1 above)
Comment (jonasfranz) [reply custom/conf/app.ini.sample:573]: @techknowlogick done

File: custom/conf/app.ini.sample [hunk 2]
```diff
@@ -570,6 +570,14 @@ MAX_RESPONSE_ITEMS = 50
...
+; Two Factor authentication with security keys
+; https://developers.yubico.com/U2F/App_ID.html
+APP_ID = https://example.com
```
Comment (lafriks) [inline custom/conf/app.ini.sample]: `app.ini.sample` should contain values that are same as default

File: public/js/index.js [hunk 3]
```diff
@@ -1432,6 +1432,133 @@ function initCodeView() {
...
 }
 
+function initU2FAuth() {
+    if($('#wait-for-key').length === 0) {
+        return
+    }
+    u2fApi.ensureSupport()
+        .then(function () {
+            $.getJSON('/user/u2f/challenge').success(function(req) {
+                console.log(req);
```
Comment (genofire) [inline public/js/index.js]: Still needed?

File: public/js/index.js [hunk 4]
```diff
@@ -1432,6 +1432,133 @@ function initCodeView() {
...
+    $.post("/user/settings/security/u2f/request_register", {
+        "_csrf": csrf,
+        "name": $('#nickname').val()
+    }).success(function(req) {
+        $("#nickname").closest("div.field").removeClass("error");
+        $('#register-device').modal('show');
+        if(req.registeredKeys === null) {
+            req.registeredKeys = []
+        }
+        console.log(req);
```
Comment (genofire) [inline public/js/index.js]: Still needed?

File: public/js/index.js [hunk 5]
```diff
@@ -1432,6 +1432,133 @@ function initCodeView() {
...
+        }).catch(function () {
+            // Fallback in case browser do not support U2F
+            window.location.href = "/user/two_factor"
+        })
+}
+function u2fSigned(resp) {
+    $.ajax({
+        url:'/user/u2f/sign',
+        type:"POST",
+        headers: {"X-Csrf-Token": csrf},
```
Comment (genofire) [inline public/js/index.js:1461]: I would like `X-CSRF-Token` correct camelCase.

File: public/js/index.js [hunk 6]
```diff
@@ -1432,6 +1432,133 @@ function initCodeView() {
...
+    }).fail(function (xhr, textStatus) {
+        u2fError(1);
+    });
+}
+
+function u2fRegistered(resp) {
+    if (checkError(resp)) {
+        return;
+    }
+    console.log(resp);
```
Comment (genofire) [inline public/js/index.js]: Still needed?

File: public/js/index.js [hunk 7]
```diff
@@ -1432,6 +1432,133 @@ function initCodeView() {
...
+    $.ajax({
+        url:'/user/settings/security/u2f/register',
+        type:"POST",
+        headers: {"X-Csrf-Token": csrf},
```
Comment (genofire) [inline public/js/index.js]: CamelCase

**Review by genofire — COMMENTED** [review]
Nice work - thank you
(Have some commants during reading)

File: public/js/index.js (hunk 5 above)
Comment (jonasfranz) [reply public/js/index.js:1461]: We use this at many other places at index.js so I would propose to use the current solution.

**Comment by lafriks** [comment]
I can not really test this but otherwise LGTM

**Comment by webjoel** [comment]
The default locale (english) for text (settings:u2f_desc): "Security keys are hardware devices containing cryptograhic keys. They could be used for two factor authentication. The security key must support the <a href="https://fidoalliance.org/">FIDO U2F</a> standard." is incorrect in word "cryptograhic", correct is "cryptographic".

**Comment by lafriks** [comment]
Somebody already fixed it

**Comment by TheAssassin** [comment]
Can you tell when this feature will be released?

**Comment by jonasfranz** [comment]
We're trying to release 1.5 in the next days since we have only ~1-2 PRs which must be merged.

**Pull request merged**